### ISSUE TITLE
Scaling the diff of positive samples

### ISSUE THREAD

**Issue by libornovax** [libornovax/master_thesis_code#6]
I discovered several things by trying to train the `macc_0.3_r2_x4` network:

Scaling the gradient
--------------------
In my loss function I average the loss over all pixel in an accumulator. Therefore, I do the same to the gradient, as shown in the code:
```
caffe_sub(count, this->_sigmoid_outputs[i]->cpu_data(), this->_accumulators[i]->cpu_data(),
          bottom[i+1]->mutable_cpu_diff());

// Scale the gradient
const Dtype loss_weight = top[0]->cpu_diff()[0] / (count / bottom[i+1]->shape(0));
caffe_scal(count, loss_weight, bottom[i+1]->mutable_cpu_diff());
```
Before, I used to have just `top[0]->cpu_diff()[0] / count` in the second to the last line, which was a problem. I concluded that the reason is that I scale the gradient down so much that in the backpropagation it basically vanishes and provides no updates, when the errors are small. Additionally, increasing learning rate in this setting to even 1.0 did not help.

However, when I scaled down the gradient a bit less (as it is in the code now) the net learned perfectly well with learning rate 0.01. The difference is I guess that when I scale down the gradient it tends to vanish faster than when the learning rate is reduced, but the gradient is larger - it makes sense because the gradient propagates as it is and only the weight update is affected by learning rate, whereas when I scaled down the gradient, the whole backpropagation was affected by the small magnitude of the gradient.

**Note** The thing that confused me was that the training loss decreased very very fast in the begining (something like 0.7, 0.1, 0.045), but then stagnated. I would say that it was because initially the error was high and thus the gradient was reasonably high, but then the error was too small to force any change of the gradient, however it was still not small enough to force the positive samples to be learned. (The error 0.045 corresponds to the error of approx. 0.045, which is already very small, but the error we need to achieve on the background samples is below 0.018=-13*ln(0.1)/2048 in order for the positive samples to start learning.)

**Summary** is that I should avoid small gradients and instead decrease learning rate, when the learning explodes.
